precompile/fee_config_manager.go

```diff
@@ -0,0 +1,268 @@
+	}
+	return feeConfig, true
+}
+
+func StoreFeeConfig(stateDB StateDB, feeConfig commontype.FeeConfig) {
+	for i := minFeeConfigFieldKey; i <= maxFeeConfigFieldKey; i++ {
+		var hashInput common.Hash
+		switch i {
+		case gasLimitKey:
+			hashInput = common.BigToHash(feeConfig.GasLimit)
+		case targetBlockRateKey:
+			hashInput = common.BigToHash(new(big.Int).SetUint64(feeConfig.TargetBlockRate))
+		case minBaseFeeKey:
+			hashInput = common.BigToHash(feeConfig.MinBaseFee)
+		case targetGasKey:
+			hashInput = common.BigToHash(feeConfig.TargetGas)
+		case baseFeeChangeDenominatorKey:
+			hashInput = common.BigToHash(feeConfig.BaseFeeChangeDenominator)
+		case minBlockGasCostKey:
+			hashInput = common.BigToHash(feeConfig.MinBlockGasCost)
+		case maxBlockGasCostKey:
+			hashInput = common.BigToHash(feeConfig.MaxBlockGasCost)
+		case blockGasCostStepKey:
+			hashInput = common.BigToHash(feeConfig.BlockGasCostStep)
+		default:
+			panic("unknown key")
+		}
+		stateDB.SetState(FeeConfigManagerAddress, common.Hash{byte(i)}, hashInput)
+	}
+}
+
+// setFeeConfig checks if the caller is permissioned for setting fee config operation.
+// The execution function parses the [input] into FeeConfig structure and sets contract storage accordingly.
+func setFeeConfig(accessibleState PrecompileAccessibleState, caller common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+	if remainingGas, err = deductGas(suppliedGas, SetFeeConfigGasCost); err != nil {
+		return nil, 0, err
+	}
+
+	if readOnly {
+		return nil, remainingGas, vmerrs.ErrWriteProtection
+	}
+
+	feeConfig, err := UnpackFeeConfigInput(input)
+	if err != nil {
+		return nil, remainingGas, err
+	}
+
+	stateDB := accessibleState.GetStateDB()
+	// Verify that the caller is in the allow list and therefore has the right to modify it
+	callerStatus := getAllowListStatus(stateDB, FeeConfigManagerAddress, caller)
+	if !callerStatus.IsEnabled() {
+		return nil, remainingGas, fmt.Errorf("%w: %s", ErrCannotChangeFee, caller)
+	}
+
+	StoreFeeConfig(stateDB, feeConfig)
+
+	// Return an empty output and the remaining gas
+	return []byte{}, remainingGas, nil
+}
+
+// getFeeConfig returns the stored fee config as an output
+func getFeeConfig(accessibleState PrecompileAccessibleState, caller common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+	if remainingGas, err = deductGas(suppliedGas, GetFeeConfigGasCost); err != nil {
+		return nil, 0, err
+	}
+
+	feeConfig, ok := GetStoredFeeConfig(accessibleState.GetStateDB())
+	if !ok {
+		// we cannot pack empty fee config, instead use zero fee config
+		output, err := PackFeeConfig(commontype.ZeroFeeConfig)
+		// should never fail to packing an empty fee config
+		if err != nil {
+			panic(err)
+		}
+		return output, remainingGas, nil
+	}
+
+	output, err := PackFeeConfig(feeConfig)
```

Comment:
nit: can we add an if condition here to return `nil, remainingGas, err` if the error is non-nil instead of returning the output regardless below?